tidy3d/plugins/autograd/invdes/penalties.py

```diff
@@ -0,0 +1,190 @@
+):
+    """Computes a penalty for erosion/dilation of a parameter map not being unity.
+
+    Accepts a parameter array normalized between 0 and 1. Uses filtering and projection methods
+    to erode and dilate the features within this array. Measures the change in the array after
+    eroding and dilating (and also dilating and eroding). Returns a penalty proportional to the
+    magnitude of this change. The amount of change under dilation and erosion is minimized if
+    the structure has large feature sizes and large radius of curvature relative to the length scale.
+
+    Parameters
+    ----------
+    filter_size : Tuple[int, ...]
+        The size of the filter to be used for erosion and dilation.
+    beta : float, optional
+        Strength of the tanh projection. Default is 100.0.
+    eta : float, optional
+        Midpoint of the tanh projection. Default is 0.5.
+    delta_eta : float, optional
+        The binarization threshold for erosion and dilation operations. Default is 0.01.
+    padding : PaddingType, optional
+        The padding type to use for the filter. Default is "reflect".
+
+    Returns
+    -------
+    Callable
+        A function that computes the erosion/dilation penalty for a given array.
+    """
+    filtproj = make_filter_and_project(filter_size, beta, eta, padding=padding)
+    eta_dilate = 0.0 + delta_eta
+    eta_eroded = 1.0 - delta_eta
+
+    def _dilate(array: np.ndarray, beta: float):
+        return filtproj(array, beta=beta, eta=eta_dilate)
+
+    def _erode(array: np.ndarray, beta: float):
+        return filtproj(array, beta=beta, eta=eta_eroded)
+
+    def _open(array: np.ndarray, beta: float):
+        return _dilate(_erode(array, beta=beta), beta=beta)
+
+    def _close(array: np.ndarray, beta: float):
+        return _erode(_dilate(array, beta=beta), beta=beta)
+
+    def _erosion_dilation_penalty(array: np.ndarray, beta: float = beta):
+        diff = _close(array, beta) - _open(array, beta)
+
+        if not np.any(diff):
+            return 0.0
+
+        return np.linalg.norm(diff) / np.sqrt(diff.size)
+
+    return _erosion_dilation_penalty
+
+
+def curvature(dp: np.ndarray, ddp: np.ndarray) -> np.ndarray:
+    """Calculate the curvature at a point given the first and second derivatives.
+
+    Parameters
+    ----------
+    dp : np.ndarray
+        The first derivative at the point, with (x, y) entries in the first dimension.
+    ddp : np.ndarray
+        The second derivative at the point, with (x, y) entries in the first dimension.
+
+    Returns
+    -------
+    np.ndarray
+        The curvature at the given point.
+
+    Notes
+    -----
+    The curvature can be positive or negative, indicating the direction of curvature.
+    The radius of curvature is defined as 1 / |κ|, where κ is the curvature.
+    """
+    num = dp[0] * ddp[1] - dp[1] * ddp[0]
+    den = np.power(dp[0] ** 2 + dp[1] ** 2, 1.5)
+    return num / den
+
+
+def bezier_with_grads(
```

Comment:
we can consider splitting the penalties into topology-based and shape-based modules?

Reply:
Yeah that makes sense, just to keep the size of these modules manageable over time. I think everything should still be available from the `tidy3d.plugins.autograd.invdes` namespace though.